File: javascript/exercises/practice/unit_test_7f181f04_1ef3_4b50_b699_b45f717c7c98/task_scheduler/.meta/task_scheduler.js
/**
 * Returns an array of task ids that can be scheduled to finish before their deadlines.
 * The returned schedule is the lexicographically smallest among all optimal schedules.
 *
 * @param {Array} tasks - List of task objects with properties:
 *                        id: number,
 *                        duration: number,
 *                        deadline: number,
 *                        dependencies: number[]
 * @returns {number[]} - Array of task ids in the order they should be executed.
 */
function scheduleTasks(tasks) {
  if (!tasks || tasks.length === 0) return [];

  // Build maps for tasks, graph, in-degrees.
  const taskMap = new Map();
  const graph = new Map();
  const inDegree = new Map();
  for (const task of tasks) {
    taskMap.set(task.id, task);
    inDegree.set(task.id, 0);
    graph.set(task.id, []);
  }

  // Build dependency graph: for each dependency, count in-degree and record outgoing edges.
  for (const task of tasks) {
    for (const dep of task.dependencies) {
      // If dependency doesn't exist in tasks, ignore (or treat as unschedulable)
      if (!taskMap.has(dep)) continue;
      graph.get(dep).push(task.id);
      inDegree.set(task.id, inDegree.get(task.id) + 1);
    }
  }

  // Use Kahn's algorithm with lexicographical order (smallest id first) to get topological order.
  const available = [];
  for (const [id, deg] of inDegree.entries()) {
    if (deg === 0) {
      available.push(id);
    }
  }
  // Sort available tasks by id (ensuring lexicographically smallest order)
  available.sort((a, b) => a - b);
  const topoOrder = [];
  while (available.length > 0) {
    const id = available.shift();
    topoOrder.push(id);
    for (const neighbor of graph.get(id)) {
      inDegree.set(neighbor, inDegree.get(neighbor) - 1);
      if (inDegree.get(neighbor) === 0) {
        available.push(neighbor);
      }
    }
    available.sort((a, b) => a - b);
  }

  // If topological order does not include all tasks, there is a cycle.
  if (topoOrder.length !== tasks.length) return [];

  // Scheduling: We simulate the sequential execution of tasks in topoOrder.
  // A task is only scheduled if all its dependencies have been scheduled
  // and including it does not violate its deadline.
  const scheduled = new Set();
  const result = [];
  let currentTime = 0;
  // Create a lookup for task dependencies for fast check.
  const depsLookup = new Map();
  for (const task of tasks) {
    depsLookup.set(task.id, new Set(task.dependencies));
  }

  // Process tasks in the lexicographically smallest topological order.
  for (const id of topoOrder) {
    const task = taskMap.get(id);
    // Check if all dependencies have been scheduled.
    let allDepsScheduled = true;
    for (const dep of depsLookup.get(id)) {
      if (!scheduled.has(dep)) {
        allDepsScheduled = false;
        break;
      }
    }
    if (!allDepsScheduled) {
      // If dependencies not met, skip this task.
      continue;
    }
    // Check if scheduling this task would complete before its deadline.
    if (currentTime + task.duration <= task.deadline) {
      currentTime += task.duration;
      scheduled.add(id);
      result.push(id);
    }
    // If the task cannot meet its deadline, it is skipped.
    // Its dependents will in turn be skipped because of unmet dependencies.
  }
  return result;
}

module.exports = { scheduleTasks };

File: java/exercises/practice/unit_test_877ca52d_2716_4088_82a6_2a5c4b747c92/task_scheduler/.meta/src/main/java/TaskScheduler.java
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.PriorityQueue;

public class TaskScheduler {

    public static List<Integer> scheduleTasks(int N, List<Task> tasks) {
        // Map from taskId to Task object for quick lookup
        Map<Integer, Task> taskMap = new HashMap<>();
        for (Task t : tasks) {
            taskMap.put(t.taskId, t);
        }

        // Build graph: for each task, record the tasks that depend on it.
        Map<Integer, List<Integer>> graph = new HashMap<>();
        // inDegree: count of prerequisites for each task.
        Map<Integer, Integer> inDegree = new HashMap<>();

        // Initialize graph and inDegree for each task.
        for (Task t : tasks) {
            graph.put(t.taskId, new ArrayList<>());
            inDegree.put(t.taskId, 0);
        }

        // Fill graph and inDegree values.
        for (Task t : tasks) {
            for (Integer dep : t.dependencies) {
                // dep -> t.taskId
                if (graph.containsKey(dep)) {
                    graph.get(dep).add(t.taskId);
                }
                inDegree.put(t.taskId, inDegree.get(t.taskId) + 1);
            }
        }

        // PriorityQueue for available tasks:
        // Comparator: first by deadline, then by duration, then taskId.
        PriorityQueue<Task> available = new PriorityQueue<>((a, b) -> {
            if (a.deadline != b.deadline) {
                return Integer.compare(a.deadline, b.deadline);
            } else if (a.duration != b.duration) {
                return Integer.compare(a.duration, b.duration);
            } else {
                return Integer.compare(a.taskId, b.taskId);
            }
        });

        // Add tasks with inDegree 0 into available pool.
        for (Task t : tasks) {
            if (inDegree.get(t.taskId) == 0) {
                available.add(t);
            }
        }

        List<Integer> schedule = new ArrayList<>();
        int currentTime = 0;
        // Process tasks in order, selecting the one with smallest deadline from the available tasks.
        while (!available.isEmpty()) {
            Task current = available.poll();
            schedule.add(current.taskId);
            currentTime += current.duration;
            // Update all tasks that depend on the current task.
            for (Integer neighborId : graph.get(current.taskId)) {
                inDegree.put(neighborId, inDegree.get(neighborId) - 1);
                if (inDegree.get(neighborId) == 0) {
                    available.add(taskMap.get(neighborId));
                }
            }
        }

        // If we haven't scheduled all tasks, there's a cycle.
        if (schedule.size() != N) {
            return new ArrayList<>();
        }

        return schedule;
    }
}